### DISCUSSION TITLE
Inner element of a template being evaluated even though x-if should be false

### DISCUSSION THREAD

**Discussion by GalahadXVI** [alpinejs/alpine#3824] [7. Bugs (archived)]
I have this following code:

```
    <template x-if="$store.active_action.state !== null && $store.active_action.state.type !== null && $store.active_action.state.type">
     <div>
        <template x-if="$store.active_action.state.show_refresh_url">
           <div>
                <template x-if="$store.active_action.state.refresh_url">
                    <form :action="$store.active_action.state.refresh_url">
                    </form>
                </template>
             </div>
         </template>
    </div>
    </template>
```

The original state is this:

```
{
 active_action:{
   state: null,
   }
  }
  ```
  
  Then it gets filled as an object
  
```
  {
   active_action:{
     state: {
       type: 'example',
       show_refresh_url: true,
       refresh_url: '/url',
       ...
     },
    }
  }
```

Later, it may return back to null again 
  
 ```
{
 active_action:{
   state: null,
   }
  }
```

When it returns back to null, I get the following error:

```
Alpine Expression Error: Cannot read properties of null (reading 'refresh_url')

Expression: "$store.active_action.state.refresh_url"

```

However, the error only seems to appear once it returns to null _after multiple times_. So if I return it back to null once, then all is good. But if I do it again, the error appears.

I can fix it by checking `active_action.state` in the second template like this:

`<template x-if="$store.active_action.state && $store.active_action.state.refresh_url">`

or this

`<template x-if="$store.active_action.state?.refresh_url">`

The question is, why is this even necessary? why is this being evaluated when the condition in the first template shouldn't pass when it returns back to null?

**UPDATE**: I've realised that removing the second template tag fixes the issue? 

This works:

```
    <template x-if="$store.active_action.state !== null && $store.active_action.state.type !== null && $store.active_action.state.type">
     <div>
             <template x-if="$store.active_action.state.refresh_url">
                 <form :action="$store.active_action.state.refresh_url">
                 </form>
             </template>
    </div>
    </template>
```

This can be easily reproduced. Check the browser console after switching back and forth between null and hydrate:
https://codepen.io/GalahadXVI/pen/MWLypoJ

I also tested this on 2.8.2 and it works fine. This is an issue that appears in Alpine 3.

**Comment by ekwoka** [comment]
I'm not positive the real reason behind why, but basically, the cloned element still exists for a tick after the x-if evaluate false, so all of its dependent values run before it's cleaned up

**Comment by Vijaykv5** [comment]
@GalahadXVI The guess error arises due to the asynchronous nature of the application. When the condition in the first template is checked, the value becomes null. Consequently, attempting to access properties of a null value triggers the error. 

By incorporating checks like **"active_action.state &&**" or using the safe navigation operator "**?.**" in the second template, you ensure the evaluation only proceeds if the object exists. This might works I believe

**Reply by GalahadXVI** [reply]
I suppose that makes sense but why would it only happen after the 2nd time it returns back to null? If I return it back to null once, it works fine.

**Reply by Vijaykv5** [reply]
Not sure on that. It's possible that the initial rendering creates some internal state or initializes certain variables, which are subsequently affected by the timing of the null value reassignment

**Comment by GalahadXVI** [comment]
I have saw similar reports in the past (#2540 #2296 #1837 #1917)

Apparently there was a fix at #2556.

Still wondering if this is a bug or expected behaviour as I'm using the latest version.

**Reply by Vijaykv5** [reply]
Haah That's great @GalahadXVI

**Reply by ekwoka** [reply]
It certainly looks like it should, but it doesn't 🤔 

My best guess is now that bit is running before those other effects have been triggered somehow...

or they are queued to run first for some reason...

**Comment by GalahadXVI** [comment]
I've figured out the cause of it and I believe it is a bug unless I'm missing something.

The actual code I'm using has another template x-if surrounding it. I didn't bother putting it in the original post because I thought it didn't matter but apparently it does.

So, this works:

```
    <template x-if="$store.active_action.state !== null && $store.active_action.state.type !== null && $store.active_action.state.type">
     <div>
         <template x-if="$store.active_action.state.refresh_url">
             <form :action="$store.active_action.state.refresh_url">
             </form>
         </template>
    </div>
    </template>
```

But this doesn't work:

```
    <template x-if="$store.active_action.state !== null && $store.active_action.state.type !== null && $store.active_action.state.type">
     <div>
        <template x-if="true">
          <div>
              <template x-if="$store.active_action.state.refresh_url">
                  <form :action="$store.active_action.state.refresh_url">
                  </form>
              </template>
           </div>
         </template>
    </div>
    </template>
```

Very odd. Surely this is a bug?

I'll update the OP.

**Reply by ekwoka** [reply]
Well, if the outer template dequeues the inner templates cleanup....

but also you shouldn't have templates immediately around templates...

**Reply by GalahadXVI** [reply]
Apologies - I'm not well versed in front end development. Would you consider this expected behaviour?

**Reply by ekwoka** [reply]
It is probably a bug that could be fixed...

I think just inline destroying the tree, instead of letting the mutation observer do it later would solve it

**Comment by GalahadXVI** [comment]
This is driving me crazy because I'm having to trawl through thousands of lines of code to "fix" the issue (i.e, I'm checking the object every time a attribute is accessed which is resulting in extremely messy code)

For example, I've had to change this:
`<span x-text="$store.active_action_5.state.refresh_url"> 
`
to this:

`<span x-text="$store.active_action && $store.active_action.state && $store.active_action.state.refresh && $store.active_action.state.refresh_url">`

This can be easily reproduced. I've created a codepen:

[https://codepen.io/GalahadXVI/pen/MWLypoJ](https://codepen.io/GalahadXVI/pen/MWLypoJ)

**Reply by ekwoka** [reply]
Well, you could just do `$store.active_action?.state?.refresh?.refresh_url`

though I assume only one of those is actually able to be null. So put the ? only after that one.

That should make this more of a minor issue.

**Reply by GalahadXVI** [reply]
No because we primarily target mobile devices and a good proportion of devices don't support optional chaining. I've already had a similar headache in the past with null coalescing.

**Reply by ekwoka** [reply]
https://caniuse.com/mdn-javascript_operators_optional_chaining

That's not true, but okay. It's been 3 years since the last mobile browser added support, and that one wasn't even getting much use then.

Your PR is welcome.

**Reply by GalahadXVI** [reply]
Just because it is green on caniuse.com doesn't mean it's not going to cause any issues. This is a genuine problem for older devices or foreign devices that don't default their webview rendering engines to chromium. This can make up a significant amount of users depending on the target audience. I'm sure theres a reason Babel is so popular...

I've already stated I'm not well versed in front end development so I'm not capable of making a PR. That doesn't mean this isn't worth mentioning.
